### DISCUSSION TITLE
Clade X pango lineage crosswalk

### DISCUSSION THREAD

**Discussion by tomasleon** [reichlab/variant-nowcast-hub#101] [Q&A] [answered]
Hi all,

Is there a handy dynamic clade X pango lineage crosswalk file that exists somewhere? We are doing our variant nowcasting by pango lineage and would have to re-classify into clades to submit to this hub. I am hoping this exists but can't seem to find it.

Thanks,
Tomas (CADPH)

**Comment by tomasleon** [comment]
@nickreich @elray1 not sure how visible discussions are to folks but hoping maybe you can triage to someone who could answer this?

**Reply by nickreich** [reply]
We are looking into this. Thanks for the ping.

**Comment by nickreich** [comment]
Hi @tomasleon -
We talked about this as a group today (including some folks from CFA) and here were our thoughts:

1. We don't really think that there is a good, clean way to map pango lineages to clades. This is inherently a noisy, not 1-to-1 mapping. Doing some post-processing to an existing model that operates on pango lineages is possible but likely would be a source of additional error/variation in the model that might be undesirable. That said, changing an existing modeling pipeline may be even less desirable.
2. We note that in the large NextStrain-curated metadata files that [we use to establish the "ground truth" for the variant hub](https://github.com/reichlab/variant-nowcast-hub?tab=readme-ov-file#clade-list) there exist, for each sequence, a pango lineage column and a nextstrain clade column. We could imagine that some analysis/cross-tabulation of these data could provide, at the current moment in time, a picture of which lineages map to which clades. There wasn't a sense among our group that this would be a process that we would recommend, but it does seem to be a possible pathway towards seeing how these different lineage/clade "calls" overlap.
3. I've reached out to the Bedford Lab that works on the NextStrain data, as they have a way of mapping clades to lineages (see, e.g. [their global sars-cov-2 variant forecast page](https://nextstrain.org/sars-cov-2/forecasts)), maybe they have a method that could be adapted for your purposes.

**Reply by rneher** [reply]
there is definitely a well defined way to collapse pango-lineages to clades, at least for all clades since 2022. In this tree

https://next.nextstrain.org/nextclade/sars-cov-2

each node is assigned both a Pango lineage and a nextstrain clade and this tree contains (nearly) all Pango lineages. The results should be a one-to-many mapping from Nextstrain clade to Pango lineage.

**Reply by nickreich** [reply]
Very cool, and I am happy to be corrected by someone more expert in this than I am! 

@rneher Do you know of a "simple" downloadable resource similar to what @tomasleon asked for at the top of this thread that provides the "crosswalk" between lineages and clades?

**Reply by trvrb** [reply, marked as answer]
If you go to https://nextstrain.org/nextclade/sars-cov-2 and click on "Download Data" at the bottom of the page and then click on "Metadata (TSV)" you'll get a TSV file that has columns for both Pango lineage (`Nextclade_pango`) and Nextstrain clade (`clade_nextstrain`).

**Reply by nlinton** [reply]
Thanks, @trvrb. Is there a static URL we can use to download the metadata (tsv) file via code vs having to point and click?

**Reply by trvrb** [reply]
This particular metadata TSV is generated on-the-fly by the browser and so there's not a simple URL endpoint to get this file. However, if you want a programmatic strategy you can do the following:
```
nextstrain remote download https://nextstrain.org/nextclade/sars-cov-2/
```
This will download the JSON file used by Auspice to display tree, etc... as `nextclade_sars-cov-2.json`. This JSON file contains a `tree` with tips and each tip will have a `Nextclade_pango` and a `clade_nextstrain` attribute. It will be a little bit of work to parse the file, but hopefully not too bad. A full description of this file schema is available here: https://github.com/nextstrain/augur/blob/master/augur/data/schema-export-v2.json, but you might find it easiest to run
```
nextstrain remote download https://nextstrain.org/ncov/open/reference
```
and just look at the file.

And alternatively you could work from `tree.json` here: https://github.com/nextstrain/nextclade_data/tree/master/data/nextstrain/sars-cov-2/wuhan-hu-1/orfs. This is only for the specific-Nextclade dataset, while the above `nextstrain remote download` solution would work for any page on the site.

**Reply by afmagee42** [reply]
Hi @rneher and @trvrb, with respect to lining up Pango lineages and Nextstrain clades, I'm having a little bit of trouble squaring what I see in the Auspice tree and in the metadata tsv and hoping you can help me figure out what I'm missing and what the implications are for translating to clades.

The Auspice tree, as promised, appears to neatly and tidily link a given Pango lineage to a single Nextstrain clade. However, in the assigned clades/lineages in the metadata, the correspondence is not so clean. 

Looking at the current metadata, sequences labeled BA.2 (full stop, no sublineages) show up with up with 18 Clade labels (plus recombinant), though most BA.2 are 21L as expected over the 21L wave. EG.5.1, KP.3.1.1, and BA.2.86 show up with 6, 5, and 3 (respectively) Pango lineage assignments, though also mostly the expected ones. However, in an older metadata file from May, sequences labeled BA.2.86 are mostly labeled 24A, not 23I.

Is the interpretation here that most of these differences are some form of assignment error? (It is after all an estimate and no estimator is perfect.) Is this some artifact of using metadata from today to look farther back? (I suppose that is perhaps another example of assignment issues, given the warnings about trusting assignments too far in the past?) Something else?

**Reply by afmagee42** [reply]
Separately, it seems worth considering the potential wrinkle caused by the order of operations between aggregating and translating/mapping. One could conceivably aggregate Pango lineages across an otherwise clean Pango lineage -> Nextstrain clade mapping.

This strikes me as most likely to happen with the "other" grab-bag taxon that shows up in a number of nowcasts, usually based on frequency threshold cutoffs. The "other" obtained by first aggregating Pango lineages and then mapping to Nextstrain clades would seem to me not to be the same "other" that would be obtained directly from clades. Differences are possible in both the composition and total size. Rare (aggregated) Pango lineages are not necessarily rare Nextstrain clades, so unless one sets up a means to break the Pango-based "other" up into its constituents, I would expect Pango lineages to end up in "other" that would otherwise have been assigned to a Nextstrain clade.

**Reply by rneher** [reply]
The tree currently used to assign lineages and clades is here:

https://nextstrain.org/fetch/data.clades.nextstrain.org/v3/nextstrain/sars-cov-2/wuhan-hu-1/orfs/2024-10-17--16-48-48Z/tree.json

I would be surprising if there was a sequence that is assigned discordant clades/lineages in this tree. If you have an example, please let us know!

**Reply by breesie** [reply]
@rneher Today we noticed that the metadata file which we're using to perform our lineages to clade crosswalks (Metadata TSV) on https://nextstrain.org/nextclade/sars-cov-2 has not been updated recently, consequently clades 24I and 24H in this round have no corresponding lineages for our CA data lineage to clade mapping, would it be possible to update the metadata file sometime soon before submissions? Thank you!

**Reply by nickreich** [reply]
I spent some time today trying to recreate the "simple" labels that Nextstrain uses to equate a Nextstrain clade and a pango lineage in a way that is similar to what is shown in the plot [here](https://github.com/nextstrain/nextclade_data/tree/master/data/nextstrain/sars-cov-2/wuhan-hu-1/orfs#nextstrain-clades). The context is that we have heard from several of our public health colleagues that they think in terms of pango lineages, and that they are eager to use the clade nowcasts from this hub to inform their thinking. So we need to be able to label plots from this hub similarly as is done in that plot linked to above, as say, "24A (JN.1)", for interpretability.

I was unable to figure out a simple way to do this. I spent over an hour unsuccessfully trying to parse and find the relevant parts of the tree.json file that @trvrb pointed to above.

And I agree with @afmagee42's observation that when I downloaded and analyzed the sequences in the metadata, there was not a 1-to-1 correspondence between Nextstrain clades and pango lineages. 

After some digging around in the nextstrain repositories, I did come across [this file](https://github.com/nextstrain/ncov/blob/master/defaults/clade_display_names.yml), which seems to provide the desired simple crosswalk. It seems to have been updated recently (two weeks ago), and while I didn't take the time to look for a script, it might just be the simple thing that @breesie has been looking for?

**Reply by nlinton** [reply]
@nickreich the [clade_display_names.yml](https://github.com/nextstrain/ncov/blob/master/defaults/clade_display_names.yml) file is not sufficient for our purposes (@breesie and CADPH team). That file only shows one lineage per clade (presumably the most prevalent one). What is needed for our nowcasting purposes is a crosswalk that shows all lineages that map to a given clade. 

That being said, we do fully support a simplified crosswalk such as that one for use for display purposes. We have found the `metadata.tsv` file @trvrb pointed to sufficient for our purposes but as @breesie pointed out, it has not been updated in quite a while (since Oct 8).

**Reply by nickreich** [reply]
@nlinton Needing the full list makes sense to me. Would there be a way to parse [the "full open" datasets](https://docs.nextstrain.org/projects/ncov/en/latest/reference/remote_inputs.html#remote-inputs-open-files) (much larger than the `metadata.tsv` file that you are using now) that are updated more or less every day to see if that could give you an updated version of the crosswalk you need?

**Reply by breesie** [reply]
Hi Nick, thanks for the suggestion. We couldn't parse the full dataset due to its size, but we were able to parse the [north America subsample](https://data.nextstrain.org/files/ncov/open/north-america/metadata.tsv.xz) sample and I noticed two I was hoping someone might be able to clarify

1. It seems like some of the samples are labeled for locations outside of North America? (pictured)
![image](https://github.com/user-attachments/assets/9771055d-9193-455c-86f4-c1a821b5fc3d)

2. For each pango_lineage, there are sometimes multiple corresponding clades. For example JN.1.16.1 is sometimes assigned clade 24A, sometimes 24H. Is this because the actual finer grain genotype of the particular sample (Nextclade_pango) (LF.7) has been assigned to clade 24H? (pictured). If this is the case, I assume trying to assign pango_lineages --instead of the more finely detailed Nextclade_pangos-- will generate inconsistencies in clade assignments?

![image](https://github.com/user-attachments/assets/21419f7f-9be7-4d9c-97f8-27a444235a1a)

Here is the most recent visualization on Nextstrain, which would still seem to indicate to me that LF.7 and its parent lineage JN.1.16.1 are still apart of clade 24A.

![image](https://github.com/user-attachments/assets/df8d1617-98aa-4c0c-ab62-b93326345614)

An update of the [Nextstrain ](https://nextstrain.org/nextclade/sars-cov-2?c=Nextclade_pango) mapping of pangos to clades --for example to account for the fact that something like the Nextclade_pango LF.7 is now apart of clade 24H and is in a distinct clade from its parent lineage JN.1.16.1 which is in clade 24A-- would be very useful. Even with the North America metadata it is a subsample so it's likely we'll be missing some clade/pango combinations making it not usable for our purposes.

**Reply by nickreich** [reply]
Hi @breesie - Thanks for the note. 

Regarding point (1) I think this is maybe something that should be filed as an issue with nextstrain, as I don't see any reason why those rows should be in the North American sample!

Regarding point (2) - I think there basically is not a super clean answer to this. my understanding is that clade/lineage assignment is a messy, stochastic process, and different algorithms (e.g., whatever pango is using vs. the nextclade algorithm) may give different answers at the "tips" of the tree. But if you "zoom out" the broader classifications of JN.1.16.1 = 24A and LF.7 = 24H do represent a mapping between labels of branches of the phylogenetic tree, as Nextstrain sees them. I don't know how long it takes Nextstrain to update all of the products/visualizations. I have been tracking the clade updates (see, e.g., the one [defining 24H and 24I](https://github.com/nextstrain/ncov/pull/1158)), which seems to give an early sense of where new clades are being defined (24J is close to being merged in as well). I don't quite understand how you are using the data shown in the second tabular snapshot above. If your modeling pipeline is tied in to using pango lineages, could you just use the Nextclade_pango field as the sequence classification? It seems like that might more reliably map to the clades than the pango classifications that come out of the "official" pango algorithm.
